File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/solution.cpp
#include <string>

class Solution {
private:
    // Helper function to check if s[left..right] is a palindrome
    bool isPalindrome(const std::string& s, int left, int right) {
        while (left < right) {
            if (s[left] != s[right]) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }

public:
    int maxPalindromes(std::string s, int k) {
        int n = s.length();
        int count = 0;
        int last_end = 0; // Tracks the end of the last selected substring

        for (int i = k - 1; i < n; ++i) {
            // Check for palindrome of length k ending at i
            if (i - k + 1 >= last_end && isPalindrome(s, i - k + 1, i)) {
                count++;
                last_end = i + 1;
            }
            // Check for palindrome of length k + 1 ending at i
            else if (i - k >= last_end && isPalindrome(s, i - k, i)) {
                count++;
                last_end = i + 1;
            }
        }

        return count;
    }
};

File: 2472-maximum-number-of-non-overlapping-palindrome-substrings/solution.py
class Solution:
    def maxPalindromes(self, s: str, k: int) -> int:
        def is_palindrome(l: int, r: int) -> bool:
            while l < r:
                if s[l] != s[r]:
                    return False
                l += 1
                r -= 1
            return True

        n = len(s)
        count = 0
        last_end = 0  # Tracks the end of the last selected substring

        for i in range(k - 1, n):
            # Check for palindrome of length k ending at i
            if i - k + 1 >= last_end and is_palindrome(i - k + 1, i):
                count += 1
                last_end = i + 1
            # Check for palindrome of length k + 1 ending at i
            elif i - k >= last_end and is_palindrome(i - k, i):
                count += 1
                last_end = i + 1

        return count